Biologically-catalysed reactions in Geochemistry

WilkAndy opened the pull request:
moose_geochemdb.json modified by adding some Biomass species and Mackinawite mineral in preparation for biogeochem.

Refs #21559

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

WilkAndy committed:
moose_geochemdb.json modified by adding some Biomass species and Mackinawite mineral in preparation for biogeochem.  Refs #21559

WilkAndy commented:
@loganharbour , i'm hoping you'll review this as i go.  My hope is to get this all done within a month.  TODO:

- [ ] core code changes
- [ ] unit-test code changes so existing unit tests run with the core code changes
- [ ] nonunit-test changes so the existing tests run with the core code changes
- [ ] documentation of the core code changes
- [ ] unit tests that comprehensively test the core code changes
- [ ] nonunit tests that demonstrate some of the new features
- [ ] documentation of biogeochemistry
- [ ] fully-fledged example(s?) of batch biogeochemistry
- [ ] fully-fledged example(s?) of reactive transport with biogeochemistry

WilkAndy committed:
core code changes needed to implement biologically-catalysed kinetic reactions in geochemistry.

Refs #21559

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1111804/) on c7c21b8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21592/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21592/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3bb5876c5c3f29649f426ff5535f8a3007658acb`

WilkAndy committed:
unit tests will now compile and pass with the core code changes

Refs #21559

WilkAndy committed:
existing tests will now pass with the core code changes.

Refs #21559

WilkAndy committed:
a small amount of documentation.

Refs #21559

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1111944/) on b4fe598 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21592/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1111972/) on b4fe598 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3bb587</th>
      <th colspan=4>#21592 b4fe59</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.54%</td>
      <td>83.55%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77922</td>
      <td>77925</td>
      <td>+3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15350</td>
      <td>15347</td>
      <td>-3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21592/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21592/coverage/framework)


## Modules coverage

### Geochemistry

<table>
  <thead>
    <tr>
      <th></th>
      <th>3bb587</th>
      <th colspan=4>#21592 b4fe59</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.61%</td>
      <td>94.85%</td>
      <td>-1.76%</td>
      <td>58.51%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4811</td>
      <td>4895</td>
      <td>+84</td>
      <td>141</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>169</td>
      <td>266</td>
      <td>+97</td>
      <td>100</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21592/coverage_diff/geochemistry)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21592/coverage/geochemistry)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21592/coverage/xfem)

</details>

## Warnings

- `geochemistry` new line coverage rate 58.51% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1111972/) on b4fe598 wanted to post the following:


The following coverage requirement(s) failed:
- `geochemistry` coverage rate 94.85% is less than the required 96.0%

WilkAndy committed:
Unit tests completed:

utils/GeochemistryKineticRateCalculator tested.
utils/GeochemistrySpeciesSwapper tested.
utils/PertinantGeochemicalSystem tested.
utils/GeochemicalSystem tested.
utils/GeochemicalSolver tested.

Added moose_testdb_e_strange.json to test edge cases in PertinantGeochemicalSystem

In-code documentation completed

Refs #21559

WilkAndy commented:
Hi @loganharbour - this is ready for review!

WilkAndy commented:
Hi @loganharbour - just checking this is on your "to-do" list and you haven't silenced notifications from github.  There's no rush on this one - i just want to know whether to email you!

WilkAndy commented:
@loganharbour - i just remembered about this!!   Is it possible to review it in the next fortnight?

loganharbour commented:
Looks good to me so far. We really need to move images eventually over to the large_media submodule - maybe this is a good time for it? I've got a few files left. Still haven't looked at documentation. Is there someone you want to bother for the doco?

WilkAndy commented:
Hi @cpgr , would you mind reviewing the documentation for this PR?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1152169/) on 6c9bdfe : invalidated by @WilkAndy

rebuild website as it appears to have been deleted

cpgr commented:
> Hi @cpgr , would you mind reviewing the documentation for this PR?

Will do it today sometime

cpgr commented on modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_zoning.md:
```diff
@@ -0,0 +1,153 @@
+
+!media geochemistry/bio_zoning.mp4 caption=Results of the aquifer-zoning biogeochemistry reactive transport simulation  id=bio_zoning.mov
+
+
```
This looks so cool! I'm really impressed that you have added this capability!

WilkAndy replied:
Thank you @cpgr .    Your nice words mean a lot to me.

WilkAndy commented:
Thank you @cpgr

WilkAndy commented on modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md:
```diff
@@ -0,0 +1,109 @@
+# Simulating biogeochemistry
```
done

WilkAndy commented on modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md:
```diff
@@ -0,0 +1,109 @@
+\mathrm{methanogen} \rightleftharpoons -\mathrm{CH}_{3}\mathrm{COO}^{-} + \mathrm{CH}_{4}\mathrm{(aq)} + \mathrm{HCO}_{3}^{-} - \mathrm{H}_{2}\mathrm{O} \ .
+\end{equation}
+
+2. The `molar volume` and `molecular weight` are largely unimportant.  Users should set reasonable numeric values, and keep those value in mind when defining the kinetic rates (below), since they impact the numeric values of the rate parameters.
```
done

WilkAndy commented on modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md:
```diff
@@ -0,0 +1,109 @@
+
+Subtracting the second from the first yields [methanogen.cata] and the equilibrium constants.  For instance $160.6192 - 157.892 = 2.7272$ is the first entry.
+
+The advantage of using a kinetic species to represent a microbe is the flexibility, for instance, mortality can be represented as a kinetic rate (see the examples below).  Moreover, in many cases, it is likely that the reactants and products of the catalysed equation are at equilibrium with the aqueous solution, making this approach more suitable than the former approach.  The main disadvantage i ...
```
Yea, i agree, both with the idea of adding a python script, and that it shouldn't be part of this enormous PR.  We'll see how much biogeochem actually gets used

WilkAndy commented on modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_death.md:
```diff
@@ -0,0 +1,30 @@
+- `intrinsic_rate_constant = k/molar_mass`
+- `multiply_by_mass = true`
+- `eta = 0`
+- `direction = DEATH`.
```
Yes, that is true.  I've now mentioned this explicitly in `biogeochemistry.md`

WilkAndy commented:
@loganharbour - when this is merged, let's move all the images across to large_media.   See #22550

WilkAndy commented:
@loganharbour - i believe this is ready for a merge now !!

WilkAndy commented:
Reminding @loganharbour to merge this PR

loganharbour reviewed (changes requested):
Please move the new content into large_media (rewrite history for this), otherwise, good for me. We should work on simplifying the utility classes into not having so many function parameters in the future :D

WilkAndy commented:
@loganharbour , boohoo - i failed with the git somehow.  i hope you can help - see #22611

WilkAndy committed:
doc and tests and cpgr changes.  Refs #21559

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 88bc8647f9f9954554485d010d7a6311e20fca71
+Subproject commit 50849f069720666d8979e0083bd19dea5dc69e11
```
**Caution!** This contains a submodule update

WilkAndy replied:
this is deliberate !

WilkAndy commented:
Oh my goodness, this passes!   Can you merge, @loganharbour ?

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 889a6e9e9bf06bdbe39220a528fc8f905b9c06df
+Subproject commit 50849f069720666d8979e0083bd19dea5dc69e11
```
**Caution!** This contains a submodule update

WilkAndy replied:
outdated - i don't need this anymore as someone else has updated large_media

rpodgorney commented:
@WilkAndy @loganharbour was the move to large media completed?

We need to move this one along ASAP please.......

WilkAndy commented:
> @WilkAndy @loganharbour was the move to large media completed?
> 
> We need to move this one along ASAP please.......

Yep, confirming the move to large media has been performed

rpodgorney commented:
Thanks @loganharbour

WilkAndy commented:
Thanks @loganharbour
